### PR TITLE
testing/opensmtpd-extras: new aport

### PR DESCRIPTION
**Pull request by Shizmob**
This aport packages the standard OpenSMTPD extras, providing it
with various filters, queues, schedulers, and tables.

### CODE REVIEW

File: testing/opensmtpd-extras/APKBUILD [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+# Maintainer: Shiz <hi@shiz.me>
+pkgname=opensmtpd-extras
+pkgver=201703132115
+pkgrel=0
+pkgdesc="OpenSMTPD addons"
+url="https://opensmtpd.org/"
+arch="all"
+license="ISC"
+depends="libevent"
+depends_filter_monkey="libressl"
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: All custom global variables (i.e. variables not specified by abuild) should be prefixed with an underscore (i.e. `_depends_filter_monkey`).

File: testing/opensmtpd-extras/APKBUILD [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+depends_table_redis="hiredis"
+depends_table_sqlite="sqlite-libs"
+makedepends="libressl-dev libevent-dev"
+makedepends_queue_python="python2-dev"
+makedepends_scheduler_python="python2-dev"
+makedepends_table_mysql="mariadb-dev"
+makedepends_table_postgres="postgresql-dev"
+makedepends_table_python="python2"
+makedepends_table_redis="hiredis-dev"
+makedepends_table_sqlite="sqlite-dev"
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: These `makedepends_*` are quite an overkill. `makedepends` specifies dependencies that should be installed for building the abuild as whole. You can’t build just a single subackage and these dependencies are not written in APKINDEX.

Just put all build dependencies into `makedepends` and don’t overcomplicate it.

File: testing/opensmtpd-extras/APKBUILD [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	filter-void
+	table-sqlite
+	table-redis
+	table-socketmap
+	table-passwd
+"
+
+for _extra in $_extras; do
+	_extraident="${_extra/-/_}"
+	_extramakedeps="$(eval "echo \$makedepends_$_extraident")"
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: It’s unnecessary to declare yet another global variable `_extraident` just to use it only one line below. You can do:

```diff
- _extramakedeps="$(eval "echo \$makedepends_$_extraident")"
+ _extramakedeps="$(eval "echo \$makedepends_${_extra/-/_}")"
```

Also you probably want `${_extra//-/_}` (replace all occurrences of `-`, not just the first one).

File: testing/opensmtpd-extras/APKBUILD [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	subpackages="$subpackages opensmtpd-$_extra:_package_extra"
+	makedepends="$makedepends $_extramakedeps"
+	case "$_extradocs" in *$_extra*) subpackages="$subpackages opensmtpd-$_extra-doc:_package_extra_doc" ;; esac
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: We don’t usually provide separate -doc subpackage for every subpackage, that’s quite an overkill. I’d just declare single -doc subpackage for all docs.

Anyway, please don’t write `case` like this on single line.

File: testing/opensmtpd-extras/APKBUILD [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	cd "$builddir"
+	local with_extras
+	local extra; for extra in $_extras; do
+		with_extras="$with_extras --with-$extra"
+	done
+
+	./configure \
+		--prefix=/usr \
+		--sysconfdir=/etc/smtpd \
+		--with-libssl=/usr/include/openssl \
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: Is it really needed to explicitly declare location of openssl headers? This is usually autodetected by autohells.

File: testing/opensmtpd-extras/APKBUILD [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+		--mandir=/usr/share/man \
+		--libexecdir=/usr/lib \
+		--with-pie \
+		--with-mantype=man \
+		$with_extras \
+		|| return 1
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: It’s not needed to use `|| return 1` anymore for new aports.

File: testing/opensmtpd-extras/APKBUILD [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+	make REDIS_CPPFLAGS='-I /usr/include/hiredis' || return 1
+}
+
+package() {
+	cd "$builddir"
+	make DESTDIR="$pkgdir" install || return 1
+}
+
+_package_extra() {
+	local extraname; extraname="${subpkgname#opensmtpd-}"
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: This is unnecessary verbose, just write `local extraname="${subpkgname#opensmtpd-}"`.

File: testing/opensmtpd-extras/APKBUILD [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	local extraident; extraident="${extraname/-/_}"
+	local extradeps; extradeps="$(eval "echo \$depends_$extraident")"
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: Same as above, `extraident` is unnecessary:

```diff
-	local extraident; extraident="${extraname/-/_}"
-	local extradeps; extradeps="$(eval "echo \$depends_$extraident")"
+	local extradeps="$(eval "echo \$depends_${extra//-/_}")"
```

File: testing/opensmtpd-extras/APKBUILD [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+	pkgdesc="OpenSMTPD $extraname addon"
+	depends="$depends $extradeps"
+	install_if="$pkgname=$pkgver-r$pkgrel"
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: This doesn’t make much sense for me, what behaviour do you expect here?

File: testing/opensmtpd-extras/remove-decls.patch [hunk 10]
```diff
@@ -0,0 +1,15 @@
+--- opensmtpd-extras-201703132115/extras/tables/table-ldap/ber.h
++++ opensmtpd-extras-201703132115-patched/extras/tables/table-ldap/ber.h
```
Comment (jirutka) [inline testing/opensmtpd-extras/remove-decls.patch]: Please replace `opensmtpd-extras-201703132115` with `a` and `opensmtpd-extras-201703132115-patched` with `b` and write a description why is this patch here.

File: testing/opensmtpd-extras/APKBUILD (hunk 5 above)
Comment (Shizmob) [inline testing/opensmtpd-extras/APKBUILD]: I copied it from the OpenSMTPD abuild -- I'll check to be sure.

File: testing/opensmtpd-extras/APKBUILD (hunk 6 above)
Comment (Shizmob) [inline testing/opensmtpd-extras/APKBUILD]: Gotcha!

File: testing/opensmtpd-extras/APKBUILD (hunk 7 above)
Comment (Shizmob) [inline testing/opensmtpd-extras/APKBUILD]: This doesn't work for all shells, though.

**Comment by Shizmob** [comment]
Thanks for the review. Updated!

File: testing/opensmtpd-extras/APKBUILD (hunk 7 above)
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: `local` itself doesn’t work in all shells, it’s not defined by POSIX. However, all common shells (ash, dash, ZSH, Bash) support it and none of them has problem with `local foo=bar`. I don’t know about any shell that supports `local`, but not with assignment.

File: testing/opensmtpd-extras/APKBUILD [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+license="ISC"
+depends=""
+_depends_filter_monkey="libressl"
+_depends_queue_python="python2"
+_depends_scheduler_python="python2"
+_depends_table_mysql="mariadb-client-libs"
+_depends_table_postgres="libpq"
+_depends_table_python="python2"
+_depends_table_redis="hiredis"
+_depends_table_sqlite="sqlite-libs"
```
Comment (jirutka) [inline testing/opensmtpd-extras/APKBUILD]: According to the [build log](https://travis-ci.org/alpinelinux/aports/builds/218695592#L1268) all these dependencies are auto-detected via dynamic links, so you can remove all these `_depends` and so simplify the abuild even more. :)

```
>>> opensmtpd-filter-monkey*: Tracing dependencies...
    libressl
    so:libc.musl-x86_64.so.1
    so:libcrypto.so.38
    so:libevent-2.1.so.6

>>> opensmtpd-queue-python*: Tracing dependencies...
    python2
    so:libc.musl-x86_64.so.1
    so:libcrypto.so.38
    so:libpython2.7.so.1.0

>>> opensmtpd-scheduler-python*: Tracing dependencies...
    python2
    so:libc.musl-x86_64.so.1
    so:libpython2.7.so.1.0

>>> opensmtpd-table-mysql*: Tracing dependencies...
    mariadb-client-libs
    so:libc.musl-x86_64.so.1
    so:libmysqlclient.so.18

>>> opensmtpd-table-postgres*: Tracing dependencies...
    libpq
    so:libc.musl-x86_64.so.1
    so:libpq.so.5

>>> opensmtpd-table-python*: Tracing dependencies...
    python2
    so:libc.musl-x86_64.so.1
    so:libpython2.7.so.1.0

>>> opensmtpd-table-redis*: Tracing dependencies...
    hiredis
    so:libc.musl-x86_64.so.1
    so:libhiredis.so.0.13

>>> opensmtpd-table-sqlite*: Tracing dependencies...
    sqlite-libs
    so:libc.musl-x86_64.so.1
    so:libsqlite3.so.0
```

**Pull request closed without merging**